### DISCUSSION TITLE
Docker image that will replace local installation.

### DISCUSSION THREAD

**Discussion by WolfgangFahl** [apache/jena#1721] [General]
I have been using the script below quite succesfully in the past years to get a command line setup that will install Jena tdbloader, fuseki, download a dataset and get everything working in one run. 
Now i have a new server which needs to run Java 8 for legacy reasons and i don't want to fiddle with newer Java versions see
https://wiki.bitplan.com/index.php/Java8

Instead i'd love to move to a docker setup and happily found
https://jena.apache.org/documentation/fuseki2/fuseki-docker.html. After trying it out i was all suprised that it will only install fuseki with no UI and no tdbloader.
https://stackoverflow.com/questions/66320295/not-able-to-create-or-access-datasets-in-jena-fuseki-docker-container pointed out part of the problem.

The images at https://hub.docker.com/r/stain/jena-fuseki seem to be no good alternative since https://github.com/stain/jena-docker/issues shows that the tdbloader problem is also not solved there.

I'd love to see a proper docker compose solution that will bring up an environment as outlined in the script below. Should i start a new project for this or is there a project out there that this could be integrated with?

```bash
#!/bin/bash
# WF 2020-05-10

# global settings
jena=apache-jena-4.4.0
tgz=$jena.tar.gz
#mirror=https://downloads.apache.org/jena/binaries
mirror=https://archive.apache.org/dist/jena/binaries
jenaurl=$mirror/$tgz
base=/hd/seel/gnd
#base=/hd/torterra/gnd
#base=/hd/luxio/gnd
data=$base/data
tdbloader=$jena/bin/tdb2.tdbloader

getjena() {
# download
if [ ! -f $tgz ]
then
  echo "downloading $tgz from $jenaurl"
  wget $jenaurl
else
  echo "$tgz already downloaded"
fi
# unpack
if [ ! -d $jena ]
then
  echo "unpacking $jena from $tgz"
  tar xvzf $tgz
else
  echo "$jena already unpacked"
fi
# create data directory
if [ ! -d $data ]
then
  echo "creating $data directory"
  mkdir -p $data
else
  echo "$data directory already created"
fi
}

#
# show the given timestamp
#
timestamp() {
 local msg="$1"
 local ts=$(date -u +"%Y-%m-%dT%H:%M:%SZ")
 echo "$msg at $ts"
}

#
# load data for the given data dir and input
#
loaddata() {
  local data="$1"
  local input="$2"
  timestamp "start loading $input to $data"
  $tdbloader --loc "$data" "$input" > tdb2-$phase-out.log 2> tdb2-$phase-err.log
  timestamp "finished loading $input to $data"
}

getjena
export TMPDIR=$base/tmp
if [ ! -d $TMPDIR ]
then
  echo "creating temporary directory $TMPDIR"
  mkdir $TMPDIR
else
  echo "using temporary directory $TMPDIR"
fi
for d in kongress geografikum
do
  file=authorities-${d}_lds.ttl
  if [ ! -f $file ]
  then
    wget https://data.dnb.de/opendata/$file.gz
    gunzip $file.gz
  else
    echo "$file already downloaded"
  fi
  loaddata $data $file 
done
```

**Comment by afs** [comment]
Try taking the jena-docker Dockerfile and scripts, and changing it to use [jena-fuseki-fulljar-4.7.0.jar](https://repo1.maven.org/maven2/org/apache/jena/jena-fuseki-fulljar/4.7.0/jena-fuseki-fulljar-4.7.0.jar).

A PR with a tested `Dockerfile-UI` to also go in jena-docker would be good to have.

**Comment by WolfgangFahl** [comment]
@afs thanks for the swift answer. I am experimenting with https://github.com/WolfgangFahl/jena-docker that started with a copy of the original files and uses github actions to try things out.
Does the fulljar include the tdbloader or would i have to download that separately?

**Reply by kinow** [reply]
That jena-docker image is new to me. I thought Andy was talking about this one: https://github.com/stain/jena-docker/

**Comment by WolfgangFahl** [comment]
```bash
docker run -i --rm -p "3030:3030" --name GndServer -t fuseki --mem /ds
[2023-01-21 15:39:20] ERROR Server          :: Can't find resourceBase (tried webapp, target/webapp, /fuseki/./webapp and /fuseki/./target/webapp)
[2023-01-21 15:39:20] ERROR Server          :: Failed to start`
```

**Comment by WolfgangFahl** [comment]
I started from https://repo1.maven.org/maven2/org/apache/jena/jena-fuseki-docker/4.7.0/ and am strugggling with the minimal approach of the Dockerfile. There is not even a jar tool available to inspect things in that image. I don't think it is necessary to be so scarce with resources these days - developer and debugging time it mostly much more precious.

**Comment by afs** [comment]
If you want all the items in the downloadable Fuseki packaging, then pull in the apache-jena-fuseki zip and work the that.

**Comment by WolfgangFahl** [comment]
For the time being i decide to not go the docker route yet. I spent too much time on debugging and getting frustrated today and have a deadline next week. I might get back to this in a few weeks.

**Comment by WolfgangFahl** [comment]
I have installed java 11 and my old script needed only minor changes to work.

**Comment by namedgraph** [comment]
Can’t you use tdbloader on host and then start the Docker container with a
mounted data folder?

BTW we have Fuseki image but it’s just using the default Dockerfile.
https://github.com/AtomGraph/fuseki-docker

**Reply by afs** [reply]
Yes, if java11 is available. Just set JAVA_HOME.

It is possible to have 2+Java on one machine.

If the restriction is installation, well, Java8 is EOL and no security updates for commercial use.

**Reply by WolfgangFahl** [reply]
tdbloader would have to be the one that needs docker but i doubt that this will be efficient since the filesystem driver of docker IMHO used to be very slow and even somewhat broken in certain configurations such as MacOS. I wonder whether this has changed in the meantime.
